### DISCUSSION TITLE
Blade pitching for below rated turbulent speed.

### DISCUSSION THREAD

**Discussion by QusayHawari** [NatLabRockies/ROSCO#120] [Q&A]
Hi,

I am running a closed loop response for the 15MW floater using a turbulent mean speed of 14m/s. 

I set the following in Servodyn:

5 = PCMode, 
5 = VSContrl,
0 = HSSBrMode
0 = YCMode.

Below you can see the wind speed [m/s] and the response of  blade 1 pitch [degs]. Occasionally, the wind speed dips below rated. In that case the blade pitch should attempt to drop to 0 degs, but looking at the pitching, it seems that it's trying to maintain the blade pitch at around 10 degs, which is the operating point at 14m/s. Is this accurate?    

![bld_pit_turb_14](https://user-images.githubusercontent.com/91909348/160183331-88c5615e-ee37-4d2c-a706-f3bde4ff70bb.PNG)

Thanks in advance,
Qusay.

**Comment by dzalkind** [comment]
This looks about right for the steady-state values.  Not sure why you're getting the 1 deg oscillations, but it seems like you were expecting that?

**Comment by QusayHawari** [comment]
Hi,

Thank you for your response.

Yes, the 1 deg oscillation is because I set 5 m of wave elevation to the structure.

My question was that since the wind speed drops below the rated speed, shouldn’t the blade pitch try to dip down a little or approach zero to switch to region 2 /2.5. I can see that the average of the pitch is exactly the operating point value at the mean of the turbulent speed.


Thanks,
Qusay.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by dzalkind** [reply]
The wind speed only drops below rated for a short period of time, and you are plotting the hub height wind speed. The rotor average wind speed `RtVAvgxh` might be a better representation of the wind the turbine/controller responds to. Either way, the pitch is determined based on the generator speed, and if it's below rated, the pitch will drop as you expect until the generator speed is above rated.  

[Some details on the theory.](https://wes.copernicus.org/articles/7/53/2022/wes-7-53-2022.html)

**Reply by QusayHawari** [reply]
Hi,

Thank you for the clarification.

I plotted the hub height wind speed and the averaged rotor wind speed for a turbulent wind input (mean of 20m/s). The figure is shown below:

![wshh_wsrot](https://user-images.githubusercontent.com/91909348/161035890-5d10f028-4e27-4e35-a77d-132763dfbfc3.jpg)

I can see that due to the 5 m wave elevations that there are some oscillations around the mean value. 

Just to clarify, in Fig.5 of the document you sent, the ' Rotor Averaged Wind Speed' is RtVAvgxh? In that case I should tune the wind speed filters more so that it matches with Wind1VelX?

Thanks,
Qusay.

**Reply by dzalkind** [reply]
Correct, if you mean Fig. 6.  RtVAvgxh is the rotor average wind speed, and `Wind1VelX` is the wind speed at the first location specified in the InflowWind input (usually hub height).  I'm not sure what you mean by the wind speed filters.

**Reply by QusayHawari** [reply]
Great thanks!

i was talking about the Kalman filter estimator in the DISCON file.  I don’t think tuning it would help the wind speed shown above (in red) to behave similarly as the  hub height wind speed as this speed is not something physical is a sense, it’s just a relative speed.


Basically , the average rotor speed is the relative speed that the rotor sees, or in other words that the controller responds to?

thanks ,
Qusay.

**Reply by dzalkind** [reply]
Correct, and since the wind speed estimator uses turbine sensors to estimate the wind speed, it is also measuring the relative wind speed as sensed by the rotor.

**Reply by QusayHawari** [reply]
Thanks,

What would you recommend I do to obtain a relative wind speed that is close to the hub height speed as shown in Fig.6 of the ROSCO documentation? Shall I generate a 'less aggressive' turbulent speed?

Qusay.

**Reply by QusayHawari** [reply]
Hi @dzalkind, 

I upgraded to Openfast v3.1 and ROSCO v2.5.0 by using the new format for OpenFAST input files in https://github.com/IEAWindTask37/IEA-15-240-RWT/tree/master/OpenFAST/IEA-15-240-RWT-UMaineSemi , and in https://github.com/NREL/ROSCO/tree/main/Test_Cases/IEA-15-240-RWT-UMaineSemi. Thank you for including the capability to tune the wind speed estimator LP filter (F_WECornerFreq).

Just to make sure that I am setting the correct closed loop operation. 

First, the setting in ServoDyn (lines 77-95) are shown in the screenshot below (for the U-Maine 15MW model):
![image](https://user-images.githubusercontent.com/91909348/161814811-7c88acc5-b565-49c3-a66c-3debc5c7e709.png)

PCMode = VSContrl = 5 and used the complied libdiscon.dll (ROSCO v 2.5.0) and openfast_x64.exe (v.3.1) found in https://github.com/NREL/ROSCO/releases and https://github.com/OpenFAST/openfast/releases ,respectively. 

There's a bit of confusion as I am not sure if GAIN_OM found on line 90 in ServoDyn is the same as VS_Rgn2K found on line 62 in DISCON.IN. Similarly for Ptch_Min/PC_MinPit , Ptch_Max/PC_MinPit , PtchRate_Min/PC_MinRat , PtchRate_Max/PC_MaxRat , Ptch_SetPnt/PC_FinePit , GenTrq_Dem/VS_RtTq. 

I used the same settings for the DISCON.IN file found in the above links except for WE_Mode = 0 to tune F_WECornerFreq, if I am not mistaken.

~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~

For the stimulation settings, I am using a turbulent wind speed with a mean of 10.5m/s on the U-Maine 15MW model.  I am also using wave elevations of 5m with a period of 10s with PotMod = ExctnMod = RdtnMod = 1.

Below are the figures of the hub height wind speed, average relative wind speed, blade 1 pitch and generator speed. 

![rot_avg_hub_ht](https://user-images.githubusercontent.com/91909348/161818968-f90c5e5b-ee90-4548-9117-2b10f99da0d5.jpg)
![rot_avg_hub_ht_zoom](https://user-images.githubusercontent.com/91909348/161818982-052511fd-7caf-432d-b667-2d79d81afff4.jpg)
![gen_spd_bld1_pit](https://user-images.githubusercontent.com/91909348/161818987-b2612762-84f8-4c1a-803e-f63b80be99c9.jpg)


The averaged relative speed is averaging near the mean value of the hub height turbulence input. The blade pitch seems to closley follow that wind speed (relative) as it dips below rated by pitching to 0 degs as seen in the zoomed in figure. As you mesntioned that the blade pitch is depends on the generator speed. If the speed is below rated then the blade pitch drops down to 0.

Would you say this is the correct operation and settings of the ROSCO controller? 


Thanks in advance,
 Qusay.

**Reply by dzalkind** [reply]
Hi Qusay,

This looks okay.  Actually, at these wind speeds, the blade pitch is going to be saturated by the peak shaver's minimum pitch limit, which is determined using the wind speed estimate (if `PS_Mode` > 0).  The pitch signal seems to have a pretty high frequency, which means that the `F_WECornerFreq` might be too high.  We typically use and test ROSCO with `WE_Mode` of 2, so if you find a better value for `F_WECornerFreq` that would be helpful.  Any reason why you are not using the wind speed estimator?  Your blade pitch also looks rate-limited in the middle plot and I'm not sure why that's different than the others.

Best, Dan

**Reply by QusayHawari** [reply]
Hi Dan,

Thanks for your quick response.

I checked the operation for WE_Mode = 2 and compared that with WE_Mode = 0. In both cases PS_Mode = 1 for pitch saturation. All other settings are the same.

![we_0_2_comps](https://user-images.githubusercontent.com/91909348/161828603-ce64e928-8314-4a5d-bd04-b7cb8ac9da9f.jpg)

I think with WE_Mode = 0, the blade pitching relates better to the wind speed as it attempts to drop when the relative speed drops below rated, whereas when setting WE_mode = 2, the blade pitch is attempting to be fixed at the operating pitch point of 10.5m/s (mean of turbulent input).


Thanks,
Qusay.

**Reply by dzalkind** [reply]
When WE_Mode = 0, ROSCO is using the hub height wind speed, which might not be indicative of the rotor average wind speed (and the estimated wind speed).  We typically only use WE_Mode = 0 for debugging the torque controller or to isolate issues in the wind speed estimator.

**Reply by QusayHawari** [reply]
Thanks, 

Yes, I see, so for the correct closed loop operation WE_Mode should be set to 2 (EKF)?

Thanks,

**Reply by dzalkind** [reply]
That is correct.

**Reply by QusayHawari** [reply]
Great!

Thank you so much.

Qusay.
